vernac/opaques.ml

```diff
@@ -0,0 +1,154 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+(************************************************************************)
+(** {6 Tables of opaque proof terms} *)
+
+(** We now store opaque proof terms apart from the rest of the environment.
+    This way, we can quickly load a first half of a .vo file without these opaque
+    terms, and access them only when a specific command (e.g. Print or
+    Print Assumptions) needs it. *)
+
+type 'a const_entry_body = 'a Entries.proof_output Future.computation
+
+type opaque_result =
+| OpaqueCertif of Safe_typing.opaque_certificate Future.computation
+| OpaqueValue of Opaqueproof.opaque_proofterm
+
+(** Current table of opaque terms *)
+
+module Summary =
+struct
+  type t = opaque_result Opaqueproof.HandleMap.t
+  let state : t ref = ref Opaqueproof.HandleMap.empty
+  let init () = state := Opaqueproof.HandleMap.empty
+  let freeze ~marshallable =
+    if marshallable then
+      let iter _ pf = match pf with
+      | OpaqueCertif cert -> ignore (Future.join cert)
+      | OpaqueValue _ -> ()
+      in
+      let () = Opaqueproof.HandleMap.iter iter !state in
+      !state
+    else !state
+  let unfreeze s = state := s
+
+  let join ?(except=Future.UUIDSet.empty) () =
+    let iter i pf = match pf with
+    | OpaqueValue _ -> ()
+    | OpaqueCertif cert ->
+      if Future.UUIDSet.mem (Future.uuid cert) except then ()
+      else
+        let cert0 = Future.join cert in
+        let () = Global.fill_opaque cert0 in
+        state := Opaqueproof.HandleMap.add i (OpaqueCertif cert) !state
+    in
+    Opaqueproof.HandleMap.iter iter !state
+
+end
+
+type opaque_disk = Opaqueproof.opaque_proofterm option array
+
+let get_opaque_disk i t =
+  let i = Opaqueproof.repr_handle i in
+  let () = assert (0 <= i && i < Array.length t) in
+  t.(i)
+
+let set_opaque_disk i (c, priv) t =
+  let i = Opaqueproof.repr_handle i in
+  let () = assert (0 <= i && i < Array.length t) in
+  let () = assert (Option.is_empty t.(i)) in
+  let c = Constr.hcons c in
+  t.(i) <- Some (c, priv)
+
+let current_opaques = Summary.state
+
+let declare_defined_opaque i (body : Safe_typing.private_constants const_entry_body) =
+  let proof = Future.chain body begin fun (body, eff) ->
+      Safe_typing.check_opaque (Global.safe_env ()) i (body, eff)
```

Comment:
> Can you explain why you want Global.safe_env to be called when the thunk is forced

I actually wondered about that when writing it. It shouldn't change anything, because the set of futures of the environment at that time will be an extension of the one at definition time. In the current phrasing, we get an anomaly if somebody tries to check an opaque after it has been filled, which is an additional safety. That's why I preferred it that way.

> Also defined_opaque seems like an oxymoron... maybe use proved instead?

Unfortunately, "defined" is quite polysemic. Not sure "proved" is any better, but I do need a name to make the distinction with the private case...

Reply:
The problem is that defined recalls `Defined.` which is not opaque.

About the safe_env, if it changes anything, then passing it explicitly would avoid any question.
My problem with the current code is that you need to understand how the table of thunks is synchronized w.r.t. Global in order to see that "it changes nothing". IMO it would be a better API to explicitly say in which env one wants the promise to be checked.

Reply:
> My problem with the current code is that you need to understand how the table of thunks is synchronized

Actually, it's much worse than that. You need to understand the STM at all to know the invariants and it takes a naughty pleasure to silently remove the rug under your feet. If anything I'd rather have a purely functional interface but unfortunately this requires changes that are way beyond this PR. There are very weird things going on but we're now entering the STM twilight zone.

Reply:
OK, so pass the safe env explicitly. In that case "which env is used to typecheck" does not depend on when the thunk is forced.

Reply:
As argued above, w.r.t. to invariants I believe that the current situation is better. Furthermore passing the env explicitly would keep a useless pointer to an old environment for no good reason. Also, it makes the interface looks like it's purely functional, but it's not...